//==============================================================================
//                                                                            
//                       Cho sanghyun's Game Classes II
//                         Ver 9.0 / Release 2017.01.03
//                                                                            
//                        Socket Classes Tutorial (Ex8)
//                                     protobuf
//-----------------------------------------------------------------------------
//
// - protobuf message definitions
//
//
//
//
//

//-----------------------------------------------------------------------------
// Declaration
//-----------------------------------------------------------------------------
// 1) Protobuf¿« Version
syntax = "proto3";

// 2) Package∏Ì
package		test;


//-----------------------------------------------------------------------------
// Definitions of 
//-----------------------------------------------------------------------------
// 1) java_declaration
option java_package			 = "com.example.tutorial";
option java_outer_classname	 = "SimpleProtos";

// 2) C# declaration
option csharp_namespace		 = "Google.Protobuf.Examples.tutorial";



//-----------------------------------------------------------------------------
// Enum
//-----------------------------------------------------------------------------
//enum eMESSAGE_COMMAND
//{
//	MESSAGE_FIRST	 = 0;
//	MESSAGE_SEND_A	 = 20;
//	MESSAGE_SEND_B	 = 21;
//	MESSAGE_SEND_C	 = 22;
//}


//-----------------------------------------------------------------------------
// Message definitions
//-----------------------------------------------------------------------------
// [Message] benchmark 01
message protobuf_message_primitive_sint32
{
	sint32				value = 1;
}

message protobuf_message_primitive_uint32
{
	uint32				value = 1;
}

message protobuf_message_primitive_sint64
{
	sint64				value = 1;
}

message protobuf_message_primitive_uint64
{
	uint64				value = 1;
}

message protobuf_message_primitive_float
{
	float				value = 1;
}

message protobuf_message_primitive_double
{
	double				value = 1;
}


// [Message] benchmark 02
message protobuf_message_string
{
	bytes				value_1 = 1;
	bytes				value_2 = 2;
	bytes				value_3 = 3;
	bytes				value_4 = 4;
	bytes				value_5 = 5;
	bytes				value_6 = 6;
	bytes				value_7 = 7;
	bytes				value_8 = 8;
}

// [Message] benchmark 03
message protobuf_message_list_int
{
	repeated int32		value_1 = 1;
}

// [Message] benchmark 04
message protobuf_message_list_string
{
	repeated bytes		value_1 = 1;
}

// [Message] benchmark 05
message protobuf_message_map_std_string_int
{
	map<string, int32>	value_1 = 1;
}

// [Message] benchmark 06
message protobuf_message_struct_primitive
{
	sint32				value_1 = 1;
	uint32				value_2 = 2;
	sint64				value_3 = 3;
	uint64				value_4 = 4;
	float				value_5 = 5;
	double				value_6 = 6;
}

// [Message] benchmark_07
message protobuf_message_struct_complex
{
	sint32				value_1 = 1;
	bytes				value_2 = 2;
	repeated int32		value_3 = 3;
	uint64				value_4 = 4;
	map<string, int32>	value_5 = 5;
}

// [Message] protobuf IDL
message protobuf_UNBELIEVABLE
{
	sint32				a = 1;
	sint64				b = 2;
	string				c = 3;
	repeated int32		d = 4;
	map<int32, string>	e = 5;
}
